Iterator version of impl

spease opened the pull request:
**Overview**

This implements the encode/decode functions with iterators. It also returns a `Result` if `decode` encounters a stream with invalid characters, rather than panicking, which is extremely undesirable. For instance, a GUI application reading a user-supplied text file would appear to just crash.

This is at best a WIP PR, but maybe just a grab bag. The only thing I'd say is really pretty important is returning a `Result` rather than panicking if the incoming data has unexpected characters in it.

**Panicking**

I tried to be very careful about where to use panicking. I left a few in - specifically in the `encode` function. These could probably be `.get_unchecked` (for the array access) and `String::from_ut8_unchecked` as the modulo-85 ensures that it's "impossible" (neglecting the possibility of single-bit flips due to cosmic rays or other shennanigans!) for larger indices or non-ASCII characters to be generated. My hope would be that the compiler would realize this and optimize the checks and potential panics out; if not, I would go ahead and use the `_unchecked` variants. The main reason to leave the checked variants in is if some code refactoring accidentally breaks these assumptions, they will fail very loudly. But for larger datasets especially, they might have a significant performance burden if the compiler does not optimize out the checks.

I think the `base64` crate even goes to the length of defining its own enum representing a subset of ASCII characters, I did not feel the benefits I was aware of for this approach would outweigh the extra code and types it would require.

**Optimizing**

This is probably not the ideally performant version, although the fact that the generics will get instantiated for specific types will hopefully be sufficient for the compiler to understand any benefits it can if the data source provided is consecutive. Ideally, the compiler would figure out you can unroll the loop slightly, but that may be too much of a stretch. Figuring out how to tweak the implementation for optimal performance would probably require a larger dataset and use of `criterion`.

If there are known, larger cases I would definitely test this on those to see if there's a performance regression (or improvement) before merging it in.

**impl Read**

I would also consider using `impl Read` rather than iterators, although this could also be a separate type, as the `base64` crate does it.

**no_std**

Finally this depends on `itertools` due to use of the `chunks` function. There may be a better alternative implementation - I don't like that itertools introduces an implicit dependency on `std`. I think it would be ideal if this crate could be compatible with `no_std`, though this would also mean changing `String` as the return type. I'm not sure if you could collect to an array somehow, so this might alter the function interface entirely to take a `&mut [u8]` in `#[no_std]` contexts.

spease committed:
Add iterator version

darkwyrm commented:
Thank you *very* much for the PR! Your code is a fascinating implementation -- I've only dipped my toes into the functional style here and there, and this gives me plenty to chew on and use in the future where appropriate.

I did some quickie performance testing on a 1MB buffer of random data and the current implementation takes 43ms to encode and 115ms to decode. The functional version 664ms and 799ms respectively. The test wasn't much more than using `stopwatch` on a buffer in VSCode, but the difference in results are significant enough that I'm going to cherry pick this.

I don't know how I missed handling errors beyond panicking, but I will be definitely merging those bits.

I have a couple questions, if you don't mind, so I can learn more:

- Why use `85u32.pow()` over the precalculated constants?
- What's the purpose of moving the character conversion functions into the scope of `encode()` and `decode()`?

spease commented:
Sorry for the lack of reply. Thanks for the quickie performance testing and adding criterion to the repository. I'm surprised at how much overhead the iterator combinators seem to be adding - see https://www.reddit.com/r/rust/comments/yaft60/zerocost_iterator_abstractionsnot_so_zerocost/ .

As I said before, I'm not 100% sure an iterator input is the way to go as `Read` might be more useful for the use-case I was thinking of - streaming compressed data. However, there's not any reason the crate has to stick to just one. It could have a trait that gets added onto qualifying iterators, or provide a special type in lieu of string that implements FromIterator. And I think the base64 crate implements a Reader type that can wrap the input for ergonomic reasons. And embedded, memory-sensitive, or high-reliability applications might want to have a function that can just convert from one buffer to another without invoking memory allocations (which could fail and panic, or just take time).

I've updated the branch - the encode function now seems to be faster than the original implementation while still using iterators. If you look in the thread there's a slightly faster version that can only work on a contiguous slice like the original. I might be able to eke out some more speed out of that one in another pass.

I suspect explicit SIMD would be the next thing to do to optimize it, considering it takes input in groups of 4s, then starts with the same input value and does essentially the same three operations with five different values. However I peeked and it looks like even the arch-specific SIMD functions that would be involved are marked as experimental right now, and the whole generic simd library is experimental.

It might also be worth looking at what it would take to make it work with rayon.

I haven't touched the decode function, it's still as slow as before.

To answer your questions:
- `85u32.pow()` should just be optimized out by the compiler (and from what I saw it was) and makes it more obvious where those constants came from rather than them just being magic numbers.
- Array indexing implicitly does a bounds check if the compiler isn't sure whether it can drop it or not. Because the implementation of encode modulos by 85, it's impossible for a value larger than 85 to be passed into the function. So that bounds check shouldn't ever be needed, but I wasn't sure if the compiler would be smart enough to realize that or not. With the benchmarks you added, I did see a small gain by switching from [] to get_unchecked, so I guess the compiler doesn't realize it.
- In the case of decode, it was purely because I wanted the two decode functions side-by-side while I was reimplementing it, and just moved the helper function inside the function rather than rename it to something I'd have to name it back.

I'm more apathetic about decode being in or outside of the parent function since it's safe to use on its own, but encode's helper function (if get_unchecked is used) I'd leave inside the encode function to make it clear it's an internal implementation detail. It might be more correct to mark the whole helper function as unsafe and instead use unsafe blocks around where it's used, but this would increase the amount of unsafe and make the function harder to read for no clear performance gain. However it would stick more closely to the convention that memory unsafety is fully encapsulated inside the unsafe blocks.

decode can probably be optimized with the lessons I learned based on encode, but I think I'm at a good stopping point for tonight so I'll leave that for later. It also looks really amenable to SIMD optimizations.
